### PR TITLE
time-varying elements

### PR DESCRIPTION
**Pull request by AdrienLeGuillou**
This PR will implement a mechanism for saving time-varying elements.

It will solves the following issues : #386 #478 

Roadmap

- [x] Validate the technical design
- [x] Find good names
    - use "record" instead of "log" 
- [x] Document

### CODE REVIEW

**Comment by AdrienLeGuillou** [comment]
The current design is as follow:

## The log function

a "log" function is made available (`log_val` for now)

```r
dat <- log_val(dat, at, "attribute_1", 1:5, "a")
```
this will store into the value `"a"` under the attribute `attribute_1` for the nodes 1 to 5 (positional indexes) at timestep `at`.

the inner representation is a list under `dat$logs` similar to 

```r
list(step = at, attribute = "attribute_1", indexes = 1:5, value = "a")
```

another example would be: 
```r
dat <- log_val(dat, at, "attribute_2", 1:5, LETTERS[1:5])

# similar to 
# list(step = at, attribute = "attribute_2", indexes = 1:5, value = c("A", "B", "C", "D", "E"))
```
in this second example, 5 values are stored as they are different for each index. In the first case, only one was saved for memory efficiency.

## The log2dfs function

The second part of this is a function to convert the `dat$log` list into a list of data.frame (and possibly other types of objects).

The current implementation is as follow:

```r
# Mock dat object for the example
dat <- list()
dat <- log_val(dat, 10, "attra", 1:5, "a")
dat <- log_val(dat, 11, "attra", 1:5, "b")
dat <- log_val(dat, 11, "attrb", 6:10, 1:5)
dat <- log_val(dat, 12, "attrb", 4:8, 1:5)

df_logs <- log2dfs(dat)
df_logs
```

Outputting a list of data.frames

```
$attra
   step attribute uids values
1    10     attra 1001      a
2    10     attra 1002      a
3    10     attra 1003      a
4    10     attra 1004      a
5    10     attra 1005      a
6    11     attra 1001      b
7    11     attra 1002      b
8    11     attra 1003      b
9    11     attra 1004      b
10   11     attra 1005      b

$attrb
   step attribute uids values
1    11     attrb 1006      1
2    11     attrb 1007      2
3    11     attrb 1008      3
4    11     attrb 1009      4
5    11     attrb 1010      5
6    12     attrb 1004      1
7    12     attrb 1005      2
8    12     attrb 1006      3
9    12     attrb 1007      4
10   12     attrb 1008      5
```

The conversion makes the ragged lists into clean data frames.

**Comment by AdrienLeGuillou** [comment]
after a discussion with @smjenness here are some precision.

1. the function should record scalar values specific to timestep and node. (So no function to record generic data without a referece to a node)
2. the default output format should be "long":

```
  attribute  uids `10`  `11`  `15`  `17`  `18`  `19`
  <chr>     <dbl> <chr> <chr> <chr> <chr> <chr> <chr>
1 attra      1001 a     NA    NA    b     b     b
2 attra      1002 a     NA    b     b     b     b
3 attra      1003 a     b     b     b     NA    b
4 attra      1004 a     b     b     b     NA    b
5 attra      1005 a     b     b     b     NA    b
6 attra      1006 NA    NA    b     NA    NA    NA
```

instead of 

```
   step attribute uids values
1    10     attra 1001      a
2    10     attra 1002      a
3    10     attra 1003      a
4    10     attra 1004      a
5    10     attra 1005      a
6    11     attra 1003      b
7    11     attra 1004      b
8    11     attra 1005      b
9    15     attra 1002      b
10   15     attra 1003      b
11   15     attra 1004      b
12   15     attra 1005      b
13   15     attra 1006      b
14   17     attra 1001      b
15   17     attra 1002      b
16   17     attra 1003      b
17   17     attra 1004      b
18   17     attra 1005      b
19   18     attra 1001      b
20   18     attra 1002      b
21   19     attra 1001      b
22   19     attra 1002      b
23   19     attra 1003      b
24   19     attra 1004      b
25   19     attra 1005      b
```

the easiest way to produce this format is with `tidyr::pivot_wider()` which mean adding `tidyr` as a dependency to `EpiModel`

3. the function would be called `record_value` for now but any better suggestion is welcome

File: NAMESPACE [hunk 1]
```diff
@@ -89,6 +89,7 @@ export(get_init)
 export(get_init_list)
 export(get_network)
 export(get_network_term_attr)
+export(get_node_records)
```
Comment (smjenness) [inline NAMESPACE:92]: I think this should be called: `get_attr_history`

File: NAMESPACE [hunk 2]
```diff
@@ -137,6 +138,8 @@ export(prevalence.icm)
 export(prevalence.icm.bip)
 export(prevalence.net)
 export(process_out.net)
+export(record_node_value)
```
Comment (smjenness) [inline NAMESPACE:141]: maybe: `record_attr_history`

File: R/get.R [hunk 3]
```diff
@@ -449,3 +449,49 @@ get_args <- function(formal.args, dot.args) {
...
+#'
+#' @param sims An \code{EpiModel} object of class \code{netsim}.
+#'
+#' @return A list of \code{data.frame}s, one for each "measure" recorded in the
+#' simulation by the `record_node_value` function.
+#'
+#' @details
+#' See the "Time Varying Attributes" vignette
+#'
+#' @export
```
Comment (smjenness) [inline R/get.R:463]: make a very small minimal @example

File: R/get.R [hunk 4]
```diff
@@ -449,3 +449,49 @@ get_args <- function(formal.args, dot.args) {
   }
   return(p)
 }
+
+#' @title Extract the node records from Network Simulations
```
Comment (smjenness) [inline R/get.R:453]: Use title case consistently for title

File: R/net.record.R [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+#' @param values the values to be recorded
+#'
+#' @return The Master list object
+#'
+#'
+#a Master list object of network models' @details
+#' See the "Time Varying Attributes" vignette
+#'
+#' @examples
+#' \dontrun{
```
Comment (smjenness) [inline R/net.record.R:23]: can be removed

File: R/net.record.R [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+#' }
+#'
+#' @export
+record_raw_object <- function(dat, at, label, object) {
+  if (is.null(dat[["raw.records"]])) {
+    dat[["raw.records"]] <- list()
+  }
+
+  element <- list(at, label, object)
+  names(element) <- c("step", "label", "object")
```
Comment (smjenness) [inline R/net.record.R:97]: can rename step --> time

File: R/net.record.R (hunk 5 above)
Comment (AdrienLeGuillou) [reply R/net.record.R:23]: the don't run part ? 

I opened the issue we discussed the other day about that: https://github.com/statnet/EpiModel/issues/567

File: DESCRIPTION [hunk 7]
```diff
@@ -49,7 +49,8 @@ Imports:
     RColorBrewer,
     ape,
     lazyeval,
-    ggplot2
+    ggplot2,
+    tidyr
```
Comment (smjenness) [inline DESCRIPTION:53]: Why the extra import here of tidyr?

Comment (smjenness) [reply DESCRIPTION:53]: It doesn't look like any new functions are imported in the NAMESPACE

**Review by smjenness — CHANGES_REQUESTED** [review]

**Comment by smjenness** [comment]
@AdrienLeGuillou : also, make sure you mark the check boxes at the top complete if the documentation is now complete.

File: DESCRIPTION (hunk 7 above)
Comment (AdrienLeGuillou) [reply DESCRIPTION:53]: my mistake, it was when we wanted to have the output in "wide" format. I used `tidyr::pivot_wider` to make the change. As we decided to go the "long" format instead, `tidyr` can be removed. I'll update this

**Pull request merged**